### DISCUSSION TITLE
x-sort and x-for out of sync

### DISCUSSION THREAD

**Discussion by craigballinger** [alpinejs/alpine#4192] [7. Bugs (archived)]
I've seen a couple references in other reports that the combination of x-sort and x-for are causing various forms of unexpected behavior. I'm running into what seems to be a pretty simple and clear case, so I wanted to explicitly document it and see if it's something I'm doing wrong or if this is currently a bug.

Codepen here: https://codepen.io/CraigAB/pen/WNWqyYd

When an item in the "Sortable" list is resorted, it immediately reverts to it's previous position, but it correctly maintains the new sort in the "Sorted" list. If the same item is sorted again into another position, it reverts to the last new position. It seems like everything is actually working behind the scenes, but the "Sortable" list UI is always one step behind the data.

Am I handling this incorrectly? Any current workaround?

**Comment by ekwoka** [comment]
This is likely a bug caused by the effect scheduler trying to prevent infinite loops.

**Reply by craigballinger** [reply]
Is there any way for me to incentivize a fix for this? I'm willing to put a bounty on it if necessary.

**Reply by ekwoka** [reply]
👀 Well, I do have my github sponsors set up...

**Comment by ekwoka** [comment]
I did find the problem, and wrote up a solution that is nearly ready to PR. I did end up mostly just refactoring the entire x-for, so ensuring it satisfied all tests took a moment, but It should be much faster on top of making this work.

**Reply by craigballinger** [reply]
Awesome! Very much appreciate your efforts resolving this one.

**Comment by ColonelThirtyTwo** [comment]
I'm looking into this too... They get out of sync because both sortable.js and alpine both attempt to modify the DOM and step on each other's toes, and [there doesn't appear to disable sortable.js doing it's dom manipulation](https://github.com/SortableJS/Sortable/issues/1931). Even if I try to fix them up, sometimes the components inside of the loop re-initialize and lose state, which is a no-go for me.

I'm not sure why sortable.js was chosen for the sort plugin given this deficiency. It struck me as odd that the documentation had no examples with how to use it with `x-for`, which, for me, is a pretty basic use case...

**Reply by ekwoka** [reply]
I'm right there with you. I do think there were better options, even barring writing up Alpine's own version. And that x-for should have been the obvious support test case.

But I have a working PR ready that solves it all.

**Reply by brianfeister** [reply]
@ekwoka did this PR ever get merged? I'm getting hit by this and it's painful!

**Reply by brianfeister** [reply]
Also willing to make a contribution @ekwoka, just not sure how long the PR / merge process will be

**Reply by ekwoka** [reply]
It hasn't. It's been sitting for a few months

**Comment by ddtbuilder** [comment]
Issue still exists...have to create weird 'view only' arrays for the DOM and save the real array behind the scenes. Very hackish. Hope this gets resolved

**Comment by ddtbuilder** [comment]
Welp, after a couple days of 'fighting the good fight' I have given up to a ridiculously ugly hack that performs as nasty as it looks, but I figure I should post it here to save the sanity of anyone else trying to beat these things into submission. Here is my problem and solution:

1. PROBLEM:  x-for and x-sort DON'T WORK TOGETHER NICELY. Actually they don't work together at all in any real use-case scenario. As many have stated, they both try to manage the DOM and interfere horribly. I think most people are using x-for to template an array of items and when the 'sorting' (moving) is done they need to persist the array that x-for iterates over. This is where Alpine and Sortable start fighting over who manages the DOM. In reality x-sort used with x-for should be more like a binding and less like a templating solution, but it fails because x-sort does NOT update the array order <--THIS IS THE REAL PROBLEM. This leaves the poor coders like me trying to make temp copies or duplicate objects in an attempt to save the order without interfering with the libraries.

2. SOLUTION (HACK): Note that I ended up using vanilla Sortable just for clarity although I'm sure this hack would work with the plugin too, however there are all sorts of oddities with the plugin (like the handle option not really working) so it was easier for me to debug without the plugin in the way. Essentially you need to wrap the entire x-for in an x-if directive that you can toggle to force a rebuild of the full x-for chunk. This has been the only 'simple' and reliable way to keep the DOM and the underlying array in sync. Hopefully there is a better fix for this and I'm just a stooge.

`
// html
template(x-if="list.updated")

// JS Sortable setup
Sortable.create( el, {
            handle: '.move-item-handle',
            animation: 150,
            onEnd: async function (evt) {
	          let currentPosition = evt.oldIndex;
	          let newPosition = evt.newIndex;
	          moveArrayItem(list.items, currentPosition, newPosition);
	          console.log(list.items);
	          list.updated = false; // key is to update this prop twice...once before nextTick and another after
	          Alpine.nextTick(() => {
		          list.updated = new Date();
		          list.save();
		          Alpine.initTree(el);
	          });
});

`

**Reply by JeffParkerOi** [reply]
I can't take the credit, but one simple solution from one of the other questions on here seems to be working (link at the bottom for reference).  Hopefully it'll just be a temporary measure until there's an official fix.

> Just ran into this as well. I'm setting the :key="Math.floor( Math.random() * 999999 )" to get around the issue. I know this is causing DOM elements to be created/destroyed , but not sure if there's another way around it.

https://github.com/alpinejs/alpine/discussions/4157#discussioncomment-13537722
